### ISSUE TITLE
add a way to execute arbitrary docker commands on a service container

### ISSUE THREAD

**Issue by CircleCode** [docker/compose#3564]
sometimes, one want to run `docker <command> <service-container>` where
- `<command>` is any valid docker command
- `<service-container>` is the container associated with service `<service>`

This can be done by mentally composing the name of the container based on project and service name, but this would be useful to have a way to
- either have a command giving the container name for a given service
  (previous command would then become `docker <command> $(docker-compose name <service>)`
- either have a command to relay a docker command to a service
  (previous command would then become `docker-compose relay <service> <command>`

if this is something aleady implemented, I apologize for not having found it
If this does not exists yet, what are your feelings about such a feature?

**Comment by shin-** [comment]
Hi,

Thank you for your suggestion. Personally, I think this proposal has several flaws.
1. It would work poorly if a service has been `scale`d to several containers. Do you then apply the command to all containers for that service? Select one arbitrarily? Demand the user select one (how would they know which one they mean to execute on?)
2. It appears to be a footgun where most of the actions it would allow would disturb the behavior of Compose itself. How is the program supposed to react when the user decides to do `docker-compose relay rm <service>`. What is the intent of the user in that case?
3. I believe most of the harmless commands on containers are already implemented in `docker-compose` anyway (`kill`, `pause`, `unpause`, `start`, `stop`, `rm`, `exec`, ...), although they might work in slightly different ways to account for the fact that Compose is designed to work with services.

Maybe you could cite some examples of situations where you would like such an option to exist, and we could figure out a better solution to achieve the same result.

**Comment by CircleCode** [comment]
Hi,

I understand the flaws you're outlining, and will partially answer all of them.

1. this semms aleady true for commands like `exec`, isn'it? so I suppose this should act the same way. In case there is some subtle difference I didn't noticed, then, I don't see any easy (fir the user) solution
2. even without `relay`, the user can do `docker rm <full name of the container>` that would disturb compose the same way… this seems like a feature that must be used with caution :-)
3. you're right, most of them, but sometimes with choices that does not fulfill all user expectations (more on this with my following examples)

I've seen 2 cases where I would have used this command:

* I had to do some `docker cp` (here is the [doc](https://docs.docker.com/engine/reference/commandline/cp/))
  This specific command does not seem to be implemented in `docker-compose`
* I have php containers that redirect _access logs_ to _stdout_ and _error logs_ to _stderr_. This makes me use very often commands like `docker logs -f <container_name> 1>/dev/null` to get only errors while using `docker logs -f <container_name> 2>/dev/null` to get only access, and `docker logs -f <container_name>` to get everything.
  It seems that `docker-compose logs <service>` redirects every log (stdout like stderr) to stdout, mixing my logs in a way I can not split again

In fact, I can also open 2 requests for these specific cases (and if you advice me to do this, I will), but I thought this proposition would be more open and could probabvly solve other (maybe upcoming) cases.

Last, I think you are right in the way `docker compose relay` appears like a _too esay way to make a lot of crap`… maybe the more appropriate solution is the other approach I proposed: `docker-compose name <service> [number]` returning the associated container name (with optional number when scalability is enabled). This command does not encourage people to do things like `docker-compose relay rm <service>` and since using subshells is aleady an advanced feature, this would prevent users without knowledge to use it in a bad way.

Thoughts?
